zcash_client_sqlite/src/wallet.rs

```diff
@@ -920,36 +1160,20 @@ fn subtree_scan_progress<P: consensus::Parameters>(
             .optional()?
             .flatten()
             .map(Ok)
-            .or_else(|| {
-                stmt_start_tree_size
-                    .query_row(
-                        named_params![":start_height": u32::from(birthday_height)],
-                        |row| row.get::<_, Option<u64>>(0),
-                    )
-                    .optional()
-                    .map(|opt| opt.flatten())
-                    .transpose()
-            })
+            // If we don't have an explicit birthday tree size, find something nearby.
+            .or_else(|| get_tree_size_near(birthday_height).transpose())
             .transpose()?;
 
-        // Get the note commitment tree size as of the end of the recover-until height.
+        // Get the note commitment tree size as of the start of the recover-until height.
         let recover_until_size = recover_until_height
-            .map(|end_height| {
-                stmt_start_tree_size
-                    .query_row(
-                        named_params![":start_height": u32::from(end_height + 1)],
-                        |row| row.get::<_, Option<u64>>(0),
-                    )
-                    .optional()
-                    .map(|opt| opt.flatten())
-            })
+            // Find a tree size near to the recover-until height
+            .map(get_tree_size_near)
             .transpose()?;
 
-        // Count the total outputs scanned so far on the birthday side of the
-        // recover-until height.
+        // Count the total outputs scanned so far on the birthday side of the recover-until height.
         let recovered_count = recover_until_height
```

Comment:
This is an `Option<Option<u64>>`. Are both levels of `Option` meaningful or can they be flattened?

Reply:
See https://github.com/zcash/librustzcash/pull/1564#discussion_r1796138798. Any nested optional derived from `recover_until_height` is meaningful in this way.